File: solutions/dart/circular-buffer/1/lib/circular_buffer.dart
// Devotion eternally offered to Shree DR.MDD 🌺

class EmptyBufferException implements Exception {}

class FullBufferException implements Exception {}

class CircularBuffer {
  int size;
  List<int> _queue = [];
  CircularBuffer(this.size);

  void clear() {
    _queue = [];
  }

  int read() {
    if (_queue.isEmpty) throw EmptyBufferException();
    return _queue.removeAt(0);
  }

  void write(int value, {bool force = false}) {
    if (_queue.length == size) {
      if (!force) throw FullBufferException();
      read();
    }
    _queue.add(value);
  }
}

File: solutions/scala/circular-buffer/1/src/main/scala/CircularBuffer.scala
// Dedicated to Shree DR.MDD with eternal reverence

class EmptyBufferException extends Exception {}
class FullBufferException extends Exception {}

class CircularBuffer(var capacity: Int) {
  private val container = Array.ofDim[Int](capacity)
  private var head = 0
  private var tail = 0
  private var size = 0

  def write(element: Int) = {
    if (size == capacity) throw new FullBufferException()
    container(tail) = element
    tail = (tail + 1) % capacity
    size += 1
  }

  def read(): Int = {
    if (size == 0) throw new EmptyBufferException()
    val elem = container(head)
    head = (head + 1) % capacity
    size -= 1
    elem
  }

  def overwrite(element: Int) = {
    if (size == capacity) read()
    write(element)
  }

  def clear() = {
    head = 0
    tail = 0
    size = 0
  }
}
